src/api/api_parsers.cpp

```diff
@@ -244,22 +244,24 @@ extern "C" {
         }
         scoped_ptr<cmd_context>& ctx = mk_c(c)->cmd();
         std::string s(str);
-        std::istringstream is(s);
-        ctx->set_regular_stream(ous);
-        ctx->set_diagnostic_stream(ous);
+        ctx->reset_parser_streams(s);
+
+        ous = ctx->parser_ous();
+        ctx->set_regular_stream(*ous);
+        ctx->set_diagnostic_stream(*ous);
         cmd_context::scoped_redirect _redirect(*ctx);
         try {
-            if (!parse_smt2_commands(*ctx.get(), is)) {
-                SET_ERROR_CODE(Z3_PARSER_ERROR, ous.str());
-                RETURN_Z3(mk_c(c)->mk_external_string(ous.str()));
+            if (!parse_smt2_commands(*ctx.get(), *ctx->parser_is())) {
+                SET_ERROR_CODE(Z3_PARSER_ERROR, ous->str());
+                RETURN_Z3(mk_c(c)->mk_external_string(ous->str()));
             }
         }
         catch (z3_exception& e) {
-            if (ous.str().empty()) ous << e.msg();
-            SET_ERROR_CODE(Z3_PARSER_ERROR, ous.str());
-            RETURN_Z3(mk_c(c)->mk_external_string(ous.str()));
+            if (ous->str().empty()) *ous << e.msg();
+            SET_ERROR_CODE(Z3_PARSER_ERROR, ous->str());
+            RETURN_Z3(mk_c(c)->mk_external_string(ous->str()));
         }
-        RETURN_Z3(mk_c(c)->mk_external_string(ous.str()));
-        Z3_CATCH_RETURN(mk_c(c)->mk_external_string(ous.str()));
+        RETURN_Z3(mk_c(c)->mk_external_string(ous->str()));
+        Z3_CATCH_RETURN(mk_c(c)->mk_external_string(ous?ous->str():""));
```

Comment:
when is ous nullptr?

Reply:
`ous` needs to be declared before `Z3_TRY` in order to be in scope in `Z3_CATCH_RETURN`. Doing this, however, forces the value to start initialized to `nullptr`, before it can be properly initialized after obtaining the `ctx` reference.

I'm guessing that if an exception is thrown before `ctx` is initialized, we can reach the catch block before `ous` has been initialized.

Reply:
Ah, c++ destructors are not invoked from C exceptions. This is indeed a tricky part.

Reply:
This line was reverted as a side effect of moving `ous` to the `api::context`.